File: app/src/main/java/io/hudepohl/gitjobs/util/EndlessScrollListener.java
package io.hudepohl.gitjobs.util;

import android.widget.AbsListView;

/**
 * Created by trent on 1/11/18.
 */

public abstract class EndlessScrollListener implements AbsListView.OnScrollListener {

    private int visibleThreshold = 5;
    private int currentPage = 1;
    private int previousTotalItemCount = 0;
    private boolean loading = true;
    private int startingPageIndex = 1;
    private int divisibilityTest = 50;

    @Override
    public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
        if (totalItemCount < previousTotalItemCount) {
            this.currentPage = this.startingPageIndex;
            this.previousTotalItemCount = totalItemCount;
            if (totalItemCount == 0) { this.loading = true; }
        }

        if (loading && (totalItemCount > previousTotalItemCount)) {
            loading = false;
            previousTotalItemCount = totalItemCount;
            currentPage++;
        }

        if (!loading && (firstVisibleItem + visibleItemCount + visibleThreshold) >= totalItemCount ) {
            if (totalItemCount % divisibilityTest == 0) {
                loading = onLoadMore(currentPage + 1, totalItemCount);
            }
        }
    }

    @Override
    public void onScrollStateChanged(AbsListView view, int scrollState) {

    }

    public abstract boolean onLoadMore(int page, int totalItemsCount);

}

File: app/src/main/java/io/hudepohl/gitjobs/util/EndlessScrollListener.kt
package io.hudepohl.gitjobs.util

import android.widget.AbsListView

abstract class EndlessScrollListener(
        private val unitsPerPage: Int,
        private val visibilityThreshold: Int) : AbsListView.OnScrollListener {

    private var page: Int = 0
    private var itemCount = unitsPerPage
    private var loading = false

    override fun onScroll(view: AbsListView, firstVisibleItem: Int,
                          visibleItemCount: Int, totalItemCount: Int) {

        if (loading && totalItemCount > itemCount) {
            loading = false
            itemCount = totalItemCount
        }

        if (!loading
                && totalItemCount % unitsPerPage == 0
                && firstVisibleItem + visibleItemCount + visibilityThreshold >= totalItemCount) {

            loading = onLoadMore(++page, totalItemCount)
        }
    }

    override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {

    }

    abstract fun onLoadMore(page: Int, totalItemsCount: Int): Boolean
}
